### PR TITLE
[Bug]Fix the bug data balance causes tablet loss

### LINKED ISSUE #6061: [Bug] Tablet is lost due to data balancing
**Describe the bug**

It is difficult to reproduce.

Suppose there are three BE nodes, A, B, C; there is a single copy of tablet_1, currently at node A. Due to data balancing, tablet_1 is moved between nodes

```
step1: A --> B
step2: B --> C
step3: C --> B 
```

In step2, tablet_1 is removed, but the data is not actually deleted in node B. The tablet is still retained in the be meta. tablet_report task of node B is reported to FE master, and FE finds that tablet_1 should not exist in node B after comparing the meta and issues a drop task. B receives the rpc and deletes the data, and removes it from the BE meta.

In step3, FE master sends clone task to B node, B receives the rpc, performs clone, and adds tablet_1 to BE meta after completion.

The normal flow is as follows.

tablet_lost-1

When the drop task rpc arrives at BE after the clone task rpc due to network or other reasons, the tablet just finished clone is deleted, resulting in data loss.

The abnormal flow is as follows.

tablet_lost-2

**Log**

Note the timestamp of the log.

fe.log

```
-- step2 clone finish, delete tablet in fe meta
2021-05-26 10:14:37,352 INFO (tablet scheduler|33) [TabletScheduler.deleteReplicaInternal():983] delete replica. tablet id: 11415, backend id: 10006. reason: DECOMMISSION state, force: false
2021-05-26 10:14:37,352 INFO (tablet scheduler|33) [TabletScheduler.removeTabletCtx():1180] remove the tablet 11415. because: redundant replica is deleted

-- be report, fe finds out tablet is not in meta, send drop task to be 
2021-05-26 10:15:17,809 WARN (Thread-48|104) [ReportHandler.deleteFromBackend():676] failed add to meta. tablet[11415], backend[10006]. errCode = 2, detailMessage = replica is enough[1-1]
2021-05-26 10:15:17,809 WARN (Thread-48|104) [ReportHandler.deleteFromBackend():689] delete tablet[11415 - 328326456] from backend[10006] because not found in meta

-- balancing, fe send clone task to be 
2021-05-26 10:15:17,812 INFO (tablet scheduler|33) [TabletScheduler.schedulePendingTablets():417] add clone task to agent task queue: tablet id: 11415, schema hash: 328326456, storageMedium: HDD, visible version(hash): 1-0, src backend: xxxxx, src path hash: 8876445914979651155, dest backend: 10006, dest path hash: -1651394959110222208

-- fe receives be finish task report 
2021-05-26 10:15:17,819 INFO (thrift-server-pool-38|2545) [TabletSchedCtx.finishCloneTask():878] clone finished: tablet id: 11415, status: HEALTHY, state: FINISHED, type: BALANCE. from backend: 10004, src path hash: 8876445914979651155. to backend: 10006, dest path hash: -1651394959110222208
2021-05-26 10:15:17,819 INFO (thrift-server-pool-38|2545) [TabletScheduler.removeTabletCtx():1180] remove the tablet 11415. because: finished

-- be tablet_report, be 10006 should have tablet 11415, but not have. set it as bad.
2021-05-26 10:16:17,810 ERROR (Thread-48|104) [ReportHandler.deleteFromMeta():573] backend [10006] invalid situation. tablet[11415] has few replica[1], replica num setting is [1]
2021-05-26 10:16:17,810 WARN (Thread-48|104) [ReportHandler.deleteFromMeta():602] tablet 11415 has only one replica 320661 on backend 10006 and it is lost, set it as bad
```

be.INFO

```
-- be receives clone task 
I0526 10:15:17.814332 27810 task_worker_pool.cpp:219] submitting task. type=CLONE, signature=11415
I0526 10:15:17.814347 27810 task_worker_pool.cpp:231] success to submit task. type=CLONE, signature=11415, task_count_in_queue=2
I0526 10:15:17.814397  6411 task_worker_pool.cpp:879] get clone task. signature:11415
I0526 10:15:17.814409  6411 engine_clone_task.cpp:83] clone tablet exist yet, begin to incremental clone. signature:11415, tablet_id:11415, schema_hash:328326456, committed_version:1
I0526 10:15:17.814419  6411 engine_clone_task.cpp:99] finish to calculate missed versions when clone. tablet=11415.328326456.da44ef84b34f17cd-26194d724a4cca95, committed_version=1, missed_versions_size=0

-- clone task finished.
I0526 10:15:17.814425  6411 tablet_manager.cpp:921] begin to process report tablet info.tablet_id=11415, schema_hash=328326456
I0526 10:15:17.814515  6411 engine_clone_task.cpp:274] clone get tablet info success. tablet_id:11415, schema_hash:328326456, signature:11415, version:1
I0526 10:15:17.814522  6411 task_worker_pool.cpp:900] clone success, set tablet infos.signature:11415

-- be receives drop task 
I0526 10:15:17.814952 30611 task_worker_pool.cpp:219] submitting task. type=DROP, signature=11415
I0526 10:15:17.814957 30611 task_worker_pool.cpp:231] success to submit task. type=DROP, signature=11415, task_count_in_queue=10

-- do drop task 
I0526 10:15:17.816692  6387 tablet_manager.cpp:462] begin drop tablet. tablet_id=11415, schema_hash=328326456
I0526 10:15:17.816697  6387 tablet_manager.cpp:1401] set tablet to shutdown state and remove it from memory. tablet_id=11415, schema_hash=328326456, tablet_path=/ssd2/515813927922307072/palobe/data/data/899/11415/328326456
I0526 10:15:17.816767  6387 tablet_meta_manager.cpp:113] save tablet meta , key:tabletmeta_11415_328326456 meta_size=1172

-- remove task info 
I0526 10:15:17.816998  6387 task_worker_pool.cpp:260] remove task info. type=DROP, signature=11415, queue_size=68
I0526 10:15:17.820333  6411 task_worker_pool.cpp:260] remove task info. type=CLONE, signature=11415, queue_size=1
```

**Cause Analysis**

There is no problem with FE timeline, BE should receive drop task first and then receive clone task. drop is executed after clone due to network or other reasons. And the BE meta does not record the version information of tablet. The task of the old data deleted the new data.

### PR DESCRIPTION
**Pull request by qidaye**
## Proposed changes

fix #6061 

Add version information in BE tablet meta, which is `replica_id` in FE meta.

Add `replica_id` information in `CloneTask`/`CreateReplicaTask`/`DropReplicaTask` of FE. 

There are 4 main changes in BE:

1. add `replica_id` information when creating a tablet
2. clone tablet with checking `replica_id`, and add `in_clone_mode` flag in tablet_meta
3. check version information(`replica_id`) and `in_clone_mode` before drop a tablet
4. add `replica_id` information when reporting tablet to FE

Two other issues have also been considered.

1. Do not affect other tasks. 
2. Compatibility of old and new data, with and without version information.

## Types of changes

What types of changes does your code introduce to Doris?
_Put an `x` in the boxes that apply_

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Code refactor (Modify the code structure, format the code, etc...)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have created an issue on (Fix #6061) and described the bug/feature there in detail
- [ ] Compiling and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If these changes need document changes, I have updated the document
- [ ] Any dependent changes have been merged

## Further comments

If this is a relatively large or complex change, kick off the discussion at dev@doris.apache.org by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: gensrc/thrift/AgentService.thrift [hunk 1]
```diff
@@ -85,11 +85,14 @@ struct TCreateTabletReq {
     12: optional bool is_eco_mode
     13: optional TStorageFormat storage_format
     14: optional TTabletType tablet_type
+    15: optional Types.TReplicaId replica_id
+    16: optional Types.TReplicaId base_replica_id
```
Comment (morningman) [inline gensrc/thrift/AgentService.thrift:89]: Add comment to explain `replica_id` and `base_replica_id`

File: be/src/agent/task_worker_pool.cpp [hunk 2]
```diff
@@ -425,11 +427,18 @@ void TaskWorkerPool::_drop_tablet_worker_thread_callback() {
         std::vector<string> error_msgs;
         TStatus task_status;
         string err;
+        TReplicaId replica_id = drop_tablet_req.__isset.replica_id ? drop_tablet_req.replica_id : 0;
         TabletSharedPtr dropped_tablet = StorageEngine::instance()->tablet_manager()->get_tablet(
-                drop_tablet_req.tablet_id, drop_tablet_req.schema_hash, false, &err);
+                drop_tablet_req.tablet_id, drop_tablet_req.schema_hash, replica_id, false, &err);
         if (dropped_tablet != nullptr) {
+            if (dropped_tablet->clone_mode()) {
```
Comment (morningman) [inline be/src/agent/task_worker_pool.cpp:434]: Why not just do this check in `tablet_manager()->drop_tablet`?

File: be/src/agent/task_worker_pool.cpp [hunk 3]
```diff
@@ -901,6 +910,14 @@ void TaskWorkerPool::_clone_worker_thread_callback() {
         DorisMetrics::instance()->clone_requests_total->increment(1);
         LOG(INFO) << "get clone task. signature:" << agent_task_req.signature;
 
+        TReplicaId replica_id = clone_req.__isset.replica_id ? clone_req.replica_id : 0;
+        // check tablet with the same tabletId existance, if exist, set tablet in clone mode
+        TabletSharedPtr exist_tablet = StorageEngine::instance()->tablet_manager()->get_tablet(
```
Comment (morningman) [inline be/src/agent/task_worker_pool.cpp:915]: I think we can do this in `EngineCloneTask`.

File: be/src/agent/task_worker_pool.cpp [hunk 4]
```diff
@@ -928,6 +945,14 @@ void TaskWorkerPool::_clone_worker_thread_callback() {
         task_status.__set_error_msgs(error_msgs);
         finish_task_request.__set_task_status(task_status);
 
+        // clone done, set clone mode false
```
Comment (morningman) [inline be/src/agent/task_worker_pool.cpp:948]: Do this in clone task.

**Comment by qidaye** [comment]
fix at #9971

**Pull request closed without merging**